### PR TITLE
#3: on push

### PR DESCRIPTION
**Pull request by umbarutiba**
@alexkontsevoy working, please review

### CODE REVIEW

File: core/utility/interfaces.ts [hunk 1]
```diff
@@ -28,15 +28,31 @@ export interface AddSymbolToNgModuleOptions {
...
   importFrom: string;
 }
 
+export interface AddImportToComponentOptions {
+  componentPath: Path;
+  importName: string;
+  importFrom: string;
+}
+
+export interface addChangeDetectionToComponentOptions {
```
Comment (astorozhevsky) [inline core/utility/interfaces.ts]: @umbarutiba please rename to `AddChangeDetectionToComponentOptions`

File: core/utility/ast-utils.ts [hunk 2]
```diff
@@ -88,6 +91,32 @@ function _addSymbolToNgModuleMetadata(options: AddSymbolToNgModuleMetadataOption
   };
 }
 
+function _addImportToFile(options: AddImportToFileOptions): Rule {
```
Comment (astorozhevsky) [inline core/utility/ast-utils.ts]: @umbarutiba are you sure we need the prefix `_` in the name of the method?

File: core/utility/ast-utils.ts [hunk 3]
```diff
@@ -152,26 +181,44 @@ export function addImportToNgModuleMetadata(options: AddSymbolToNgModuleOptions)
...
-
-    const change = insertImport(source, options.modulePath, options.importName, options.importFrom);
+    const source = ts.createSourceFile(options.componentPath, sourceText, ts.ScriptTarget.Latest, true);
+    const insertProperty = ',\n  changeDetection: ChangeDetectionStrategy.OnPush';
 
-    if (change instanceof InsertChange) {
-      const recorder = host.beginUpdate(options.modulePath);
-      recorder.insertLeft(change.pos, change.toAdd);
-      host.commitUpdate(recorder);
+    const classDecorator = findNodes(source, ts.SyntaxKind.Decorator)[0];
```
Comment (astorozhevsky) [inline core/utility/ast-utils.ts]: @umbarutiba if `findNodes` will return the empty array or null then here will be the error.

**Review by astorozhevsky — CHANGES_REQUESTED** [review]

File: core/utility/ast-utils.ts (hunk 2 above)
Comment (umbarutiba) [reply core/utility/ast-utils.ts]: другие не экспортируемые  утилиты имеют такой префикс  _addSymbolToNgModuleMetadata()  
_buildRoute()

File: core/utility/ast-utils.ts (hunk 3 above)
Comment (umbarutiba) [reply core/utility/ast-utils.ts]: const classDecorator = findNodes(source, ts.SyntaxKind.Decorator).find((item) => item.getText().includes('@Component'));
    if (!classDecorator) {
      throw new SchematicsException(`File ${options.componentPath} does not have a decorator.`);
    }

**Comment by umbarutiba** [comment]
@astorozhevsky  - fixed, please check

File: core/utility/ast-utils.ts [hunk 4]
```diff
@@ -152,26 +181,44 @@ export function addImportToNgModuleMetadata(options: AddSymbolToNgModuleOptions)
...
 
 export function addImportToModule(options: AddImportToModuleOptions): Rule {
+  return addImportToFile({
+    filePath: options.modulePath,
+    importName: options.importName,
+    importFrom: options.importFrom
+  });
+}
+
+export function addImportToComponent(options: AddImportToComponentOptions): Rule {
```
Comment (dmitryusenko) [inline core/utility/ast-utils.ts]: @umbarutiba I am not sure if we need `addImportToComponent` and `addImportToModule` wrappers as they are doing the same thing now. Can't we call `addImportToFile` instead whenever we want to add import to component or a module?

File: core/utility/ast-utils.ts [hunk 5]
```diff
@@ -152,26 +181,44 @@ export function addImportToNgModuleMetadata(options: AddSymbolToNgModuleOptions)
...
 
-    if (change instanceof InsertChange) {
-      const recorder = host.beginUpdate(options.modulePath);
-      recorder.insertLeft(change.pos, change.toAdd);
-      host.commitUpdate(recorder);
+    const classDecorator = findNodes(source, ts.SyntaxKind.Decorator).find((item) => item.getText().includes('@Component'));
+    if (!classDecorator) {
+      throw new SchematicsException(`File ${options.componentPath} does not have a decorator.`);
     }
+    const lastPropertyDecorator = findNodes(classDecorator, ts.SyntaxKind.PropertyAssignment).pop();
```
Comment (dmitryusenko) [inline core/utility/ast-utils.ts]: @umbarutiba Could you please rename this to `lastDecoratorProperty`?

File: src/shared-module/files/component/__name@dasherize__.component.ts [hunk 6]
```diff
@@ -1,8 +1,9 @@
-import { Component } from '@angular/core';
+import { Component<% if (isNgrxInstalled && hasNotRoot) { %>, ChangeDetectionStrategy<% } %> } from '@angular/core';
 
 @Component({
   selector: '<%= dasherize(((hasSection && !hasPage) ? (section + ' '): '') + ((hasSection && hasParentPage) ? (parentPage + ' ') : '') + ((hasSection && hasPage) ? (page + ' ') : '') + name) %>',
   templateUrl: '<%= dasherize(name) %>.html',
-  styleUrls: ['<%= dasherize(name) %>.scss']
+  styleUrls: ['<%= dasherize(name) %>.scss']<% if (isNgrxInstalled && hasNotRoot) { %>,
```
Comment (dmitryusenko) [inline src/shared-module/files/component/__name@dasherize__.component.ts:6]: @umbarutiba Could you please explain why we need to check `hasNotRoot` here? If I understood correctly, this check will prevent shared components from having `changeDetection` property in `@Component` decorator. Why shouldn't we add it for these components?

Comment (umbarutiba) [reply src/shared-module/files/component/__name@dasherize__.component.ts:6]: @dmitryusenko   
image

File: core/utility/ast-utils.ts (hunk 4 above)
Comment (umbarutiba) [reply core/utility/ast-utils.ts]: @astorozhevsky  I also think that these wrappers are not needed.

**Comment by umbarutiba** [comment]
@dmitryusenko  What do you think about [this ](https://github.com/RonasIT/angular-schematics/pull/21#discussion_r382381956)  
rest is done

File: core/utility/ast-utils.ts (hunk 4 above)
Comment (astorozhevsky) [reply core/utility/ast-utils.ts]: @umbarutiba Ok, let's move it to the single method

File: src/shared-module/files/component/__name@dasherize__.component.ts (hunk 6 above)
Comment (astorozhevsky) [reply src/shared-module/files/component/__name@dasherize__.component.ts:6]: @umbarutiba @dmitryusenko let's discuss it. Maybe all components should have onPush strategy.

**Comment by astorozhevsky** [comment]
@umbarutiba please resolve merge conflicts

**Review by astorozhevsky — APPROVED** [review]

File: src/shared-module/files/component/__name@dasherize__.component.ts (hunk 6 above)
Comment (astorozhevsky) [reply src/shared-module/files/component/__name@dasherize__.component.ts:6]: @umbarutiba @dmitryusenko decided to leave as is

**Pull request merged**